doc/fwutil/fwutil.md

```diff
@@ -462,6 +509,333 @@ Aborted!
+
+        Raises:
+            RuntimeError: auto-update failure cause
+        """
+        raise NotImplementedError
+```
+The return_code of auto_update_firmware() which indicates the firmware auto-update  status, will be logged in "fw_au_status" under "/tmp/firmwareupdate/" directory by fwutil and the status file will be used for the `fwutil show auto_update_status` command. 
+
+Any scheduled component firmware update needs to be logged in "<boot_type>_fw_au_task".
+Each vendor needs to implement the reboot auto-update task handler plugin which is platform specific plugin and is expected to be named as "platform_fw_au_reboot_handle".
+The plugin will parse the "<boot_type>_fw_au_task" file and perform the update during the reboot if needed. 
+The contents of the task file can be defined by the platform vendors and be utilized by the platform specific reboot plugin.
+
+```
+{
+    "<boot_type>": [
+        {
+            "status": <true/false>,
+            "comp": "<component_path>",
+            "info": "<installed/updated/scheduled/ErrorMsg>"
+        },
+        {
+            "status": <true/false>,
+            "comp": "<component_path>",
+            "info": "<installed/updated/scheduled/ErrorMsg>"
+        }
+    ]
+}
+```
+Here is the example of the firmware update status file with cold reboot as boot_type action.
+And the components available for the update are BIOS, CPLD, and SSD.
+In this example, BIOS firmware got updated, CPLD firmware got installed but powercycle* is needed, and SSD firmware update is scheduled during boot.
+*powercycle can be triggered by cold reboot script in this case.
+```
+admin@sonic:~/fwutil$ cat /tmp/firmwareupdate/fw_au_status
+{
+    "cold": [
+        {
+            "status": true,
+            "comp": "MSN2700/BIOS",
+            "info": "updated"
+        },
+        {
+            "status": true,
+            "comp": "MSN2700/SSD",
+            "info": "scheduled"
+        },
+        {
+            "status": true,
+            "comp": "MSN2700/CPLD1",
+            "info": "installed"
+        }
+    ]
+}
+```
+
+
+The FWutil infra section below indicates the necessary changes to add support for auto update during reboot.
+The Platform plugins section describes the plugins that a platform must implement to support auto updates when the platform component APIs are not available.
+
+Automatic FW installation requires default platform_components.json to be created and placed at:
+_sonic-buildimage/device/<platform_name>/<onie_platform>/platform_components.json_
+Recommended image path is "/lib/firmware/<vendor>".
+
+Auto-update command can support the standalone custom firmware image package with --fw-image option.
+The package can be any format between ".tar" or ".tar.gz" and should have the "platform_components.json", the firmware image(s).
+The package can have the auto-update plugin if the platform doesn't support platform api or if plugin needs to be updatedplugin.
```

Comment:
@sujinmkang IMHO, the original idea of FW package was to provide an ability to do the upgrades on platforms which don't support platform API. Do we really want to use it for upgrades which involves platform API, when fwutil is already provides an infrastructure for that (e.g, this can be achived with FW deb packages)?

Reply:
@nazariig I think it's good point. I understand we use the FW deb package to installed firmware which is missing from the current package. We can support both for the new firmware update based on the current image. But when we use a deb package to perform auto-update, we need 3 steps, push the firmware deb package, install the deb package and fwutil auto-update while this FW package approach needs 2 steps for auto update, push the FW package and fwutil aut-update with the FW package.